ssz/utils.py

```diff
@@ -1,67 +1,120 @@
 
 
-def infer_sedes(value):
-    """
-    Try to find a sedes objects suitable for a given Python object.
-    """
-    if isinstance(value.__class__, BaseSedes):
-        # Mainly used for `Serializable` Classes
-        return value.__class__
+def pack(serialized_values: Sequence[bytes]) -> Tuple[bytes]:
+    if len(serialized_values) == 0:
+        return (EMPTY_CHUNK,)
 
-    elif isinstance(value, bool):
-        return boolean
+    item_size = len(serialized_values[0])
+    items_per_chunk = get_items_per_chunk(item_size)
 
-    elif isinstance(value, int):
-        raise TypeError("uint sedes object or uint string needs to be specified for ints")
+    number_of_items = len(serialized_values)
+    number_of_chunks = (number_of_items + (items_per_chunk - 1)) // items_per_chunk
 
-    elif isinstance(value, (bytes, bytearray)):
-        return bytes_sedes
+    chunk_partitions = partition(items_per_chunk, serialized_values, pad=b"")
+    chunks_unpadded = (b"".join(chunk_partition) for chunk_partition in chunk_partitions)
 
-    elif isinstance(value, Sequence):
-        return infer_list_sedes(value)
+    full_chunks = tuple(take(number_of_chunks - 1, chunks_unpadded))
+    last_chunk = first(chunks_unpadded)
+    if len(tuple(chunks_unpadded)) > 0:
+        raise Exception("Invariant: all chunks have been taken")
 
-    elif isinstance(value, Iterable):
-        raise TypeError("Cannot infer list sedes for iterables that are not sequences")
+    return full_chunks + (last_chunk.ljust(CHUNK_SIZE, b"\x00"),)
 
-    else:
-        raise TypeError(f"Did not find sedes handling type {type(value).__name__}")
 
+def get_next_power_of_two(integer: int) -> int:
+    powers_of_two = (2**exponent for exponent in count(0))
+    greater_or_equal_powers = dropwhile(lambda power: power < integer, powers_of_two)
+    return first(greater_or_equal_powers)
 
-def get_duplicates(values):
-    counts = collections.Counter(values)
-    return tuple(
-        item
-        for item, num in counts.items()
-        if num > 1
+
+def pad_chunks(chunks: Sequence[bytes]) -> Tuple[bytes]:
+    unpadded_number_of_chunks = len(chunks)
+    padded_number_of_chunks = get_next_power_of_two(unpadded_number_of_chunks)
+    padding = (EMPTY_CHUNK,) * (padded_number_of_chunks - unpadded_number_of_chunks)
+    return tuple(chunks) + padding
+
+
+def hash_layer(child_layer: Sequence[bytes]) -> Tuple[bytes]:
+    child_pairs = partition(2, child_layer)
+    parent_layer = tuple(
+        hash_eth2(left_child + right_child)
+        for left_child, right_child in child_pairs
     )
+    return parent_layer
+
+
+def merkleize(chunks: Sequence[bytes]) -> bytes:
+    padded_chunks = pad_chunks(chunks)
+    number_of_layers = int(math.log2(len(padded_chunks))) + 1
+
+    layers = take(number_of_layers, iterate(hash_layer, padded_chunks))
+    root_layer = last(layers)
+    if not len(root_layer) == 1:
+        raise Exception("Invariant: root level of merkle tree has single element")
```

Comment:
Something I've been playing with lately for unpacking *single* value iterables with implicit validation is:

```python
root_layer, = last(layers)
```

This uses tuple unpacking and raises an exception if `last(layers)` isn't a length-1 iterable.  Saves the check but still blows up if the value is malformed.